### PR TITLE
Through the styles an override can be provided for specified road classes

### PR DESCRIPTION
**Pull request by cafesilencio**
## Description

Very often there are gaps in the traffic coverage on maps.  In the map's style there is often a substitution that is performed for "unknown" traffic conditions that results in the color for "low" traffic being substituted for "unknown" traffic conditions for certain road classes.  However the Nav. SDK didn't do such replacements and instead represented the traffic conditions received from the Directions API.  This could result in some visual inconsistency since some road sections would display the color for "unknown" traffic on the route line while the map color would display the color for "low" traffic.

The work in this PR makes available an optional substitution via the styles.xml.  

- [ ] I have added any issue links
- [ ] I have added all related labels (`bug`, `feature`, `new API(s)`, `SEMVER`, etc.)
- [ ] I have added the appropriate milestone and project boards

### Goal

For an organization that has chosen to substitute the color for "low" traffic congestion for "unknown" traffic for certain road classes on the map, there should also be a method of making the same substitution in the Nav. SDK for the route line.

### Implementation

The styles.xml has a new string array for including road classes. For any road class included in the array when a section of a route has an "unknown" traffic indication that section will be colored with the "low" traffic color rather than the color designated for "unknown".

## Screenshots or Gifs

Screenshot_20201019-163305
Screenshot_20201019-163648

## Testing

Testing can be a challenge. The road class data needed is not included in the driving/traffic profile by default only in the driving profile. However the driving profile won't contain the traffic congestion data in the route response. It's necessary to use a token from 1Tap when requesting a route in order to get the traffic and road class data using the driving/traffic profile. Also a map style that displays the traffic congestion on the map will be needed. Using the 1Tap map style and token may be helpful in this regard.

- [x] I have tested locally (including `SNAPSHOT` upstream dependencies if needed) through testapp/demo app and run all activities to avoid regressions
- [x] I have tested via a test drive, or a simulation/mock location app
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have updated the `CHANGELOG` including this PR
- [ ] We might need to update / push `api/current.txt` files after running `$> make core-update-api` (Core) / `$> make ui-update-api` (UI) if there are changes / errors we're 🆗 with (e.g. `AddedMethod` changes are marked as errors but don't break SemVer) 🚀 If there are SemVer breaking changes add the `SEMVER` label. See [Metalava](https://github.com/mapbox/mapbox-navigation-android/blob/master/docs/metalava.md) docs

### CODE REVIEW

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt [hunk 1]
```diff
@@ -1505,6 +1527,67 @@ internal class MapRouteLine(
...
+                            it.geometryIndex() != null && it.mapboxStreetsV8()?.roadClass() != null
+                        }?.toList()?.forEach {
+                            roadClassMap[it.geometryIndex()!!] =
+                                it.mapboxStreetsV8()!!.roadClass()!!
+                        }
+
+                    return route.legs()?.mapNotNull {
+                        it.annotation()?.congestion()
+                    }?.flatten()?.mapIndexed { index, congestion ->
+                        val distanceOffset = distanceList.subList(0, index).sum()
```
Comment (LukasPaczos) [inline libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt:1554]: While trying to draw a route from Wrocław to Paris, it takes ~25 seconds for it to show up while the main thread is frozen. Profiling shows that this line is consuming ~61% of the main thread's time.
Screenshot from 2020-10-26 12-03-25

Would you be able to give all of the calculations from this PR a pass and optimize them?

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt [hunk 2]
```diff
@@ -1505,6 +1527,67 @@ internal class MapRouteLine(
                 else -> y2
             }
         }
+
+        fun getRouteLineTrafficExpressionData(route: DirectionsRoute):
```
Comment (LukasPaczos) [inline libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt:1531]: Could you explain what the new methods do in a doc entry?

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt [hunk 3]
```diff
@@ -1505,6 +1527,67 @@ internal class MapRouteLine(
...
+            routeDistance: Double,
+            congestionColorProvider: (String, Boolean) -> Int,
+            isPrimaryRoute: Boolean,
+            trafficOverrideRoadClasses: List<String>
+        ): List<RouteLineExpressionData> {
+            return trafficExpressionData.map {
+                val percentDistanceTraveled = it.distanceFromOrigin / routeDistance
+                val trafficIdentifier =
+                    if (
+                        trafficOverrideRoadClasses.contains(it.roadClass ?: "unknown") &&
```
[the remark this answers is not in the document]
Comment (cafesilencio) [reply libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt:1580]: There are cases I've come across where the road class for a given intersection is null.  I want to provide a way for a developer to indicate what should be done in this case.  My intention was that "unknown" could be included in the collection of the road classes defined in the styles.xml.  This would provide an option for making a substitution when the road class is null or not making a substitution. If I make the change you suggest there would be no option and instead the same behavior would be implemented for everyone.

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt (hunk 1 above)
Comment (cafesilencio) [reply libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt]: Good catch. One bug I found was that the calculation of the traffic congestion was being done twice. So eliminating that improved things quite a bit.  I also removed the numerous calls to sum the distance of the route sections and instead keep a running total of the distance as the code iterates through the route sections.  In addition the code looks for duplicate congestion/road class entries to create a simpler expression for the gradient line. So for example if the first route section has a congestion level = low and the second route section has the same congestion level and road class I omit the second item in the gradient expression since it wouldn't change the appearance of the gradient line.  When I create a route from San Francisco to the middle of the U.S. It is taking less than 25ms on my phone.

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt (hunk 2 above)
Comment (cafesilencio) [reply libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt:1480]: done

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt [hunk 4]
```diff
@@ -1505,6 +1527,67 @@ internal class MapRouteLine(
...
+        fun getRouteLineTrafficExpressionData(route: DirectionsRoute):
+            List<RouteLineTrafficExpressionData> {
+                val distances = route.legs()?.mapNotNull {
+                    it.annotation()?.distance()
+                }?.flatten()
+                ifNonNull(distances) { distanceList ->
+                    val roadClassMap = mutableMapOf<Int, String>()
+                    route.legs()?.asSequence()
+                        ?.map { it.steps() }
+                        ?.filterNotNull()
```
[the remark this answers is not in the document]
Comment (cafesilencio) [reply libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt]: done

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt [hunk 5]
```diff
@@ -1321,7 +1340,7 @@ internal class MapRouteLine(
          */
         fun calculateRouteLineSegments(
```
[the remark this answers is not in the document]
Comment (cafesilencio) [reply libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt:1358]: done

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt [hunk 6]
```diff
@@ -1330,13 +1349,16 @@ internal class MapRouteLine(
                 ?.flatten() ?: listOf()
 
             return when (congestionSections.isEmpty()) {
-                false -> calculateRouteLineSegmentsFromCongestion(
```
[the remark this answers is not in the document]
Comment (cafesilencio) [reply libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt:1333]: done

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (LukasPaczos) [reply libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt:1566]: I think we should limit the scope of the feature to match the data that's returned by the Directions API. If the road class is unavailable, it means that it also isn't available in the map data so there will be no discrepancy in styling between the map and the route, which ultimately is what we're trying to resolve.

If we're trying to categorize unavailable road classes as `"unknown"`, I think that it should be done on the Directions API level.

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt [hunk 7]
```diff
@@ -531,24 +541,42 @@ internal class MapRouteLine(
      * @param directionsRoutes the routes to be represented on the map.
      */
     fun draw(directionsRoutes: List<DirectionsRoute>) {
-        val featureDataProvider: () -> List<RouteFeatureData> =
-            getRouteFeatureDataProvider(directionsRoutes)
-        reinitializeWithRoutes(directionsRoutes, featureDataProvider)
-        drawRoutes(routeFeatureData)
+        ThreadController.getMainScopeAndRootJob().scope.launch {
```
Comment (LukasPaczos) [inline libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt:544]: Let's avoid adding asynchronous code to the `MapRouteLine` at this point. While I think it's a good idea and we should definitely explore it, I'd rather see it done in a separate PR since there are many more considerations than just offloading the work. As an example, the current code might easily run into a race condition when `draw` is called in quick succession with 2 different parameters which will result in displaying an outdated route.

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt [hunk 8]
```diff
@@ -827,13 +849,15 @@ internal class MapRouteLine(
...
+        val start = System.currentTimeMillis()
         val segments = calculateRouteLineSegments(
             routeData.route,
-            lineString,
+            trafficBackfillRoadClasses,
             true,
             ::getRouteColorForCongestion
         )
+        val end = System.currentTimeMillis() - start
+        Timber.e("*** total time for segments = $end num segments ${segments.size}")
```
Comment (LukasPaczos) [inline libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt:860]: Remove the logs.

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt [hunk 9]
```diff
@@ -1308,10 +1341,13 @@ internal class MapRouteLine(
          * Calculates line segments based on the legs in the route line and color representation
          * of the traffic congestion. The items returned can be used to create a style expression
          * which can be used to style the route line. The styled route line will be colored
-         * according to the traffic conditions indicated in the @param route.
+         * according to the traffic conditions indicated in the @param route. For any road class
+         * included in the @param trafficBackfillRoadClasses 'unknown' traffic values will be
+         * colored with the low traffic color instead of the color configured for unknown traffic
+         * congestion.
```
Comment (LukasPaczos) [inline libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt:1347]:
```suggestion
         * according to the traffic conditions indicated in the @param route. For any road class
         * included in the @param trafficBackfillRoadClasses, all route segments with an 'unknown' traffic congestion annotation and a matching road class will be
         * colored with the low traffic color instead of the color configured for unknown traffic
         * congestion.
```

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt [hunk 10]
```diff
@@ -1505,6 +1462,118 @@ internal class MapRouteLine(
...
+                                RouteLineTrafficExpressionData(
+                                    runningDistance,
+                                    congestion,
+                                    roadClassMap[index]
+                                )
+                            )
+                        } else {
+                            runningDistance += distanceList[index - 1]
+                            if (routeLineTrafficData.last().trafficCongestionIdentifier ==
+                                congestion &&
```
Comment (LukasPaczos) [inline libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt:1515]: Can we apply the same strategy of simplifying the expression also when road classes are not available, or not used for backfilling?

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt (hunk 8 above)
Comment (cafesilencio) [reply libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt:860]: oops, thanks for pointing this out.

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (cafesilencio) [reply libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt:1566]: Fair enough. I'll make an adjustment.

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt (hunk 10 above)
Comment (cafesilencio) [reply libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt:1515]: It already works for both conditions. There are unit tests for the method using a routes with and without road classes. The test without road classes goes from 54 to 5 items in the result with that snippet of code.

File: libnavigation-ui/src/test/java/com/mapbox/navigation/ui/route/MapRouteLineTest.kt [hunk 11]
```diff
@@ -1124,18 +1112,18 @@ class MapRouteLineTest {
...
-            "[\"rgba\", 86.0, 168.0, 251.0, 1.0], 0.33272266, " +
-            "[\"rgba\", 86.0, 168.0, 251.0, 1.0], 0.39298487, " +
-            "[\"rgba\", 86.0, 168.0, 251.0, 1.0], 0.48991048, " +
-            "[\"rgba\", 255.0, 77.0, 77.0, 1.0], 0.504132, " +
-            "[\"rgba\", 86.0, 168.0, 251.0, 1.0], 0.8017851, " +
-            "[\"rgba\", 86.0, 168.0, 251.0, 1.0], 1.0000086, [\"rgba\", 86.0, 168.0, 251.0, 1.0]]"
+            "[\"rgba\", 86.0, 168.0, 251.0, 1.0], 0.05612764, " +
+            "[\"rgba\", 86.0, 168.0, 251.0, 1.0], 0.10589197, " +
+            "[\"rgba\", 86.0, 168.0, 251.0, 1.0], 0.15093622, " +
+            "[\"rgba\", 86.0, 168.0, 251.0, 1.0], 0.20669733, " +
```
Comment (LukasPaczos) [inline libnavigation-ui/src/test/java/com/mapbox/navigation/ui/route/MapRouteLineTest.kt:1118]: We should able to filter out these color duplicates, right? At this point, we are not concerned about the congestion level anymore, but just about the color we actually assign. This might significantly simplify the expression for longer routes and improve performance even more.

Let's do that as a follow-up.

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt (hunk 10 above)
Comment (LukasPaczos) [reply libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt:1500]: Cool! What I'm also looking for is removing the color duplicates from the expression itself, comment down below.

Also, which tests are impacted? I don't notice them in the diff.

File: libnavigation-ui/src/test/java/com/mapbox/navigation/ui/route/MapRouteLineTest.kt [hunk 12]
```diff
@@ -1124,18 +1112,18 @@ class MapRouteLineTest {
...
-            "[\"rgba\", 86.0, 168.0, 251.0, 1.0], 0.1905395, " +
-            "[\"rgba\", 86.0, 168.0, 251.0, 1.0], 0.26964808, " +
-            "[\"rgba\", 255.0, 77.0, 77.0, 1.0], 0.27944908, " +
-            "[\"rgba\", 86.0, 168.0, 251.0, 1.0], 0.33272266, " +
-            "[\"rgba\", 86.0, 168.0, 251.0, 1.0], 0.39298487, " +
-            "[\"rgba\", 86.0, 168.0, 251.0, 1.0], 0.48991048, " +
-            "[\"rgba\", 255.0, 77.0, 77.0, 1.0], 0.504132, " +
-            "[\"rgba\", 86.0, 168.0, 251.0, 1.0], 0.8017851, " +
-            "[\"rgba\", 86.0, 168.0, 251.0, 1.0], 1.0000086, [\"rgba\", 86.0, 168.0, 251.0, 1.0]]"
+            "[\"rgba\", 86.0, 168.0, 251.0, 1.0], 0.05612764, " +
```
Comment (LukasPaczos) [inline libnavigation-ui/src/test/java/com/mapbox/navigation/ui/route/MapRouteLineTest.kt:1115]: Why did the offset values change in the tests? They don't seem to match any previous values from the array that would indicate that we simply skipped some entries.

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt [hunk 13]
```diff
@@ -575,26 +593,19 @@ internal class MapRouteLine(
...
      */
     fun updatePrimaryRouteIndex(route: DirectionsRoute) {
-        this.primaryRoute = route
+        this@MapRouteLine.primaryRoute = route
         val partitionedRoutes = routeFeatureData.partition { it.route == primaryRoute }
         routeFeatureData.apply {
             clear()
             addAll(listOf(partitionedRoutes.first, partitionedRoutes.second).flatten())
         }
+        updateRouteTrafficSegments(routeFeatureData.first())
```
[the remark this answers is not in the document]
Comment (cafesilencio) [reply libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt:608]: This was a bug fix for the calculating of the traffic segments happening twice.

File: libnavigation-ui/src/test/java/com/mapbox/navigation/ui/route/MapRouteLineTest.kt (hunk 12 above)
Comment (cafesilencio) [reply libnavigation-ui/src/test/java/com/mapbox/navigation/ui/route/MapRouteLineTest.kt]: I misunderstood what you were asking about. Look at the tests getRouteLineTrafficExpressionDataWhenStreetClassDataExists and getRouteLineTrafficExpressionDataWhenStreetClassDoesNotExist.

There was a slight change in the way the distances in the route were added up but from what I looked at the difference was within 4 or 5 decimal places. With such a small degree of difference I don't think it's erroneous because it's a tiny fraction of a metere.

File: libnavigation-ui/src/test/java/com/mapbox/navigation/ui/route/MapRouteLineTest.kt (hunk 11 above)
Comment (cafesilencio) [reply libnavigation-ui/src/test/java/com/mapbox/navigation/ui/route/MapRouteLineTest.kt]: done

**Comment by LukasPaczos** [comment]
I pushed on https://github.com/mapbox/mapbox-navigation-android/tree/lp-backfill-test a unit test that takes a small strip of motorway and tries to backfill the color. You'll see that the test fails while the result in the `CustomRouteStylingActivity` looks like this:
device-2020-10-30-124104

even though all (besides last) intersections have a road class `motorway`, the backfill is checkered.

File: libnavigation-ui/src/test/java/com/mapbox/navigation/ui/route/MapRouteLineTest.kt (hunk 12 above)
Comment (LukasPaczos) [reply libnavigation-ui/src/test/java/com/mapbox/navigation/ui/route/MapRouteLineTest.kt]: Oh, so what you're saying is that turf calculations were different than what Directions API returns? If that's the case, I guess it makes us only more precise.

**Comment by cafesilencio** [comment]
We may have a difference in interpretation of the data. Using the route from your test:

traffic-data

There is a one to one relationship between the distance collection and the traffic congestion collection. 
So at,
index 0 distance point 3.7 there is traffic 'unknown'
index 1 distance point 12.4 (+ 3.7) there is traffic 'unknown'

There is no one to one relationship with the road classes. They may not exist and if they do they have a geometry index so the first road class that is present in the route has a geometry index of 0 which results in distance point 3.7 / traffic unknown / road class motorway. There isn't another road class until index 3. For indexes 1 and 2 there is no road class but there is a distance and traffic congestion value. Since the road class at distance/congestion indexes 1 and 2 is null that is creating the gaps you see.  For this reason I originally had included the ability to specify "unknown" in the road classes list in the styles.xml so that in cases where the road class was unknown (null) and the congestion was unknown the color could be substituted with low.

**Comment by LukasPaczos** [comment]
> We may have a difference in interpretation of the data.

Yes, that's it.

When you see a list of intersections in the response, like this:
```json
"intersections": [
  {
    "location": [
      16.983162,
      51.045254
    ],
    "bearings": [
      135
    ],
    "classes": [
      "toll",
      "motorway"
    ],
    "entry": [
      true
    ],
    "out": 0,
    "geometry_index": 0,
    "is_urban": false,
    "admin_index": 0,
    "mapbox_streets_v8": {
      "class": "motorway"
    }
  },
  {
    "location": [
      16.983438,
      51.045078
    ],
    "bearings": [
      135,
      315
    ],
    "classes": [
      "toll",
      "motorway"
    ],
    "entry": [
      true,
      false
    ],
    "in": 1,
    "out": 0,
    "geometry_index": 3,
    "is_urban": false,
    "admin_index": 0,
    "mapbox_streets_v8": {
      "class": "motorway"
    }
  },
...
```

It means that the `"class": "motorway"` spans from geometry index 0 to geometry index 3, not that indices 1 and 2 are null. We'll need to do this logical merger when we parse the data.

/cc @danpat please correct me if I'm wrong

**Comment by danpat** [comment]
@LukasPaczos Your interpretation is correct - the geometry *between* intersections is uninterrupted and doesn't change properties until the next intersection.

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt [hunk 14]
```diff
@@ -1534,6 +1476,146 @@ internal class MapRouteLine(
...
+                            )
+                        } else {
+                            if (routeLineTrafficData.last().trafficCongestionIdentifier ==
+                                congestion &&
+                                routeLineTrafficData.last().roadClass == roadClassMap[index]
+                            ) {
+                                // continue
+                            } else if (routeLineTrafficData.last().roadClass != null &&
+                                roadClassMap[index] == null
+                            ) {
```
Comment (LukasPaczos) [inline libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt:1535]: This check is too lenient. In a scenario when the congestion level changes between intersections, we will skip and ignore it.

For example, running `getRouteLineExpressionDataWithStreetClassOverrideWhenHasStreetClassesOnMotorway` test and changing the congestion array from all `unknown`:
```
"congestion":["unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown"]
```
to:
```
"congestion":["unknown","severe","severe","unknown","severe","severe","severe","severe","severe","severe","severe","severe","severe","severe","severe","severe","severe","severe","severe","severe","severe","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown","unknown"]
```
still lets the test pass. It ignores all of the severe congestion levels because they do not start on intersections geometry points.

Could you add another test case for this scenario?

Comment (cafesilencio) [reply libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt]: Added additional testing.

**Comment by LukasPaczos** [comment]
https://github.com/mapbox/mapbox-navigation-android/tree/lp-congestion-change-outside-intersection adds a failing test case. The scenario is:
1. We are on a motorway with backfill enabled.
2. Route has `severe` congestion across the board.
3. A small piece of geometry _between intersections_ has `unknown` congestion.

This piece of `unknown` traffic should be backfilled to `low` congestion, but it's not, it's drawn with `unknown` congestion color instead.

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt [hunk 15]
```diff
@@ -1588,6 +1530,159 @@ internal class MapRouteLine(
...
+                    { Expression.color(trafficColor) }
+                )
+                if (index == 0) {
+                    expressionDataToReturn.add(
+                        RouteLineExpressionData(
+                            percentDistanceTraveled,
+                            colorExpression
+                        )
+                    )
+                } else if (Expression.color(trafficColor) !=
```
Comment (LukasPaczos) [inline libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt:1673]:
```suggestion
                } else if (colorExpressionMap.get(trafficColor) !=
```

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt [hunk 16]
```diff
@@ -1588,6 +1530,159 @@ internal class MapRouteLine(
...
+                    "Route leg annotation does not contain distance and/or congestion " +
+                        "so route line traffic cannot be calculated."
+                )
+                return listOf()
+            }
+
+        private fun getRoadClassForIndex(roadClassMap: Map<Int, String>, index: Int): String? {
+            var indexCountdown = index
+            while(indexCountdown >= 0) {
+                if (roadClassMap.containsKey(indexCountdown)) {
```
Comment (LukasPaczos) [inline libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt:1616]: In order for this to be correct, you would need to add all intersection to this map, not only the ones that have road classes.

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt [hunk 17]
```diff
@@ -1588,6 +1530,159 @@ internal class MapRouteLine(
...
+                    route.legs()?.asSequence()
+                        ?.mapNotNull { it.steps() }
+                        ?.flatten()
+                        ?.mapNotNull { it.intersections() }
+                        ?.flatten()
+                        ?.filter {
+                            it.geometryIndex() != null
+                        }?.toList()?.forEach {
+                            roadClassMap[it.geometryIndex()!!] =
+                                it.mapboxStreetsV8()?.roadClass() ?: "unknown"
```
Comment (LukasPaczos) [inline libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt:1564]:
```suggestion
                                it.mapboxStreetsV8()?.roadClass() ?: "intersection_without_road_class_fallback"
```

Let's fall back to something that definitely won't make it into the actual response in the future.

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt [hunk 18]
```diff
@@ -1588,6 +1530,159 @@ internal class MapRouteLine(
...
+                Timber.i(
+                    "Route leg annotation does not contain distance and/or congestion " +
+                        "so route line traffic cannot be calculated."
+                )
+                return listOf()
+            }
+
+        private fun getRoadClassForIndex(roadClassMap: Map<Int, String>, index: Int): String? {
+            var indexCountdown = index
+            while(indexCountdown >= 0) {
```
Comment (LukasPaczos) [inline libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt:1615]:
```suggestion
            for(indexCountdown in index downTo 0) {
```

Comment (cafesilencio) [reply libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt:1615]: I couldn't remember that syntax and I was in a hurry. :)

File: libnavigation-ui/src/test/java/com/mapbox/navigation/ui/route/MapRouteLineTest.kt [hunk 19]
```diff
@@ -1775,6 +1768,264 @@ class MapRouteLineTest {
...
+                it.geometryIndex() != null && it.mapboxStreetsV8()?.roadClass() != null
+            }
+            ?.map { it.mapboxStreetsV8()!!.roadClass() }
+            ?.toList()
+
+        val result = getRouteLineTrafficExpressionData(route)
+
+        assertEquals(distances.size, result.size)
+        assertEquals(distances.first(), result[1].distanceFromOrigin, 0.0)
+        assertEquals(result[0].roadClass, roadClasses!!.first())
```
Comment (LukasPaczos) [inline libnavigation-ui/src/test/java/com/mapbox/navigation/ui/route/MapRouteLineTest.kt:1792]:
```suggestion
        assertEquals("service", result[0].roadClass)
```

and remove the local `roadClasses` which is outdated anyway

File: libnavigation-ui/src/test/java/com/mapbox/navigation/ui/route/MapRouteLineTest.kt [hunk 20]
```diff
@@ -1775,6 +1768,264 @@ class MapRouteLineTest {
...
+        assertEquals(result[2].distanceFromOrigin, distances.subList(0, 2).sum(), 0.0)
```
Comment (LukasPaczos) [inline libnavigation-ui/src/test/java/com/mapbox/navigation/ui/route/MapRouteLineTest.kt:1793]:
```suggestion
        assertEquals(distances.subList(0, 2).sum(), result[2].distanceFromOrigin, 0.0)
```

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt [hunk 21]
```diff
@@ -1588,6 +1530,157 @@ internal class MapRouteLine(
...
+            }
+
+        private fun getRoadClassForIndex(roadClassMap: Map<Int, String>, index: Int): String? {
+            for (indexCountdown in index downTo 0) {
+                if (roadClassMap.containsKey(indexCountdown)) {
+                    return roadClassMap[indexCountdown]
+                }
+            }
+            return null
+        }
```
Comment (LukasPaczos) [inline libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt:1620]: This method seems to be very inefficient and takes up more than 90% of the main thread's time:
Screenshot from 2020-11-05 13-57-30

I'm currently running into ANRs when trying to get a route from DC to LA.

I'll try to come up with some optimizations.

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt [hunk 22]
```diff
@@ -1588,6 +1530,157 @@ internal class MapRouteLine(
...
+                    route.legs()?.asSequence()
+                        ?.mapNotNull { it.steps() }
+                        ?.flatten()
+                        ?.mapNotNull { it.intersections() }
+                        ?.flatten()
+                        ?.filter {
+                            it.geometryIndex() != null
+                        }?.toList()?.forEach {
+                            roadClassMap[it.geometryIndex()!!] = it.mapboxStreetsV8()?.roadClass()
+                                ?: "intersection_without_class_fallback"
```
Comment (LukasPaczos) [inline libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt:1564]: I now discovered that the `geometryIndex` values are scoped per leg. This means that if we have a multi-leg route, the indices and class names will start to overwrite each other.

Comment (LukasPaczos) [reply libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt:1564]: https://github.com/mapbox/mapbox-navigation-android/pull/3672/commits/8ca2165d880d552b7cd2b04e87565ad973b87c43 adds the test case (currently failing).

Comment (LukasPaczos) [reply libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt:1563]: https://github.com/mapbox/mapbox-navigation-android/pull/3672/commits/40775c11927f31378c3bc40492de90eca05a5144 is a first fix attempt.

File: libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt (hunk 21 above)
Comment (LukasPaczos) [reply libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt]: Resolved with https://github.com/mapbox/mapbox-navigation-android/pull/3672/commits/db56e146905077f252201c4183a56b5a789a8483. The route still freezes the UI for ~2 seconds but it matches the experience on `master`. We should prioritize moving the operations to the worker thread with 1.3.

Comment (LukasPaczos) [reply libnavigation-ui/src/main/java/com/mapbox/navigation/ui/route/MapRouteLine.kt]: Follow up in https://github.com/mapbox/mapbox-navigation-android/issues/3737.

**Review by LukasPaczos — APPROVED** [review]
:ship: 

Great work here @cafesilencio. All of those iterations and test cases ultimately put us in a good spot. Glad we caught so many issues so early!

**Comment by cafesilencio** [comment]
> 🚢
> 
> Great work here @cafesilencio. All of those iterations and test cases ultimately put us in a good spot. Glad we caught so many issues so early!

Thanks for identifying the edge cases and contributing.
